tools/install.py

```diff
@@ -197,15 +197,56 @@ def files(action):
 def headers(action):
   def wanted_v8_headers(files_arg, dest):
     v8_headers = [
+      'deps/v8/include/cppgc/allocation.h',
```

Comment:
It might be better to go back to excluding the headers you don't want to export?

Whilst I'm here though, I'll mention that we actually do make use of the `v8-inspector.h` and `v8-inspector-protocol.h` headers, so we have a local patch to include those also.  Those were the only headers that used to be in the disallow list (See: https://github.com/nodejs/node/commit/38f32386c138073c6a020ce79085daea15e7b800).

Reply:
I think the concern is that we don't want to accidentally expand the API surface. If we have a disallow list we might fail to notice when V8 adds new headers and then breaks the API by removing them. If the list is explicit it helps us notice about changes like this and we can make some fixups or adjust the timing of the V8 upgrade.

Reply:
I don't understand the subject enough to have an opinion on whether we should expose all these additional headers, but I'm afraid that doing so will effectively block us from doing semver-minor V8 upgrades in the future.

Reply:
FYI, The v8 headers missing from this list (as of 11.3) are:

```
+      'deps/v8/include/v8-fast-api-calls.h',
+      'deps/v8/include/v8-inspector-protocol.h',
+      'deps/v8/include/v8-inspector.h',
+      'deps/v8/include/v8-metrics.h',
+      'deps/v8/include/v8-unwinder-state.h',
+      'deps/v8/include/v8-util.h',
+      'deps/v8/include/v8-value-serializer-version.h',
+      'deps/v8/include/v8-version-string.h',
+      'deps/v8/include/v8-wasm-trap-handler-posix.h',
+      'deps/v8/include/v8-wasm-trap-handler-win.h',
```

Reply:
Assuming we publish all these internal cppgc headers. Can we still update V8 in an ABI-compatible way if they make breaking changes to them (while keeping `v8-cppgc.h` without breaking changes)

Reply:
`v8-cppgc.h` does not provide everything you need.  You need to directly include a bunch of the `cppgc/` headers directly.  In our existing code, we are including the following:

```
$ git grep "#include.*cppgc" | cut -d ":" -f2 | sort -u
#include <cppgc/allocation.h>
#include <cppgc/garbage-collected.h>
#include <cppgc/name-provider.h>
#include <cppgc/persistent.h>
#include <cppgc/prefinalizer.h>
#include <v8-cppgc.h>
```

The internal headers I assume you're not supposed to directly include, but they're included by the non-internal headers, so they need to be available for transitive inclusion.

You would have to be as careful as you are now about changes to these headers, when determining whether it's safe to do a given v8 upgrade as a node minor release.

Reply:
From what I can tell, if cppgc's ABI compatibility is included in our ABI compatibility contract (no breaking changes in semver-minor), we have more changes to review when doing a V8 upgrade. Also because how the cppgc constructs are template-heavy, it's easier for the cppgc updates to break the ABI.

On the other hand, I think we could also consider excluding cppgc from our ABI compatibility contract, or at least list them as experimental so they are excluded from the contract for now until the update to the headers really slow down. Our ABI compatibility contract only really extends to V8, whereas cppgc is more like a standalone C++ GC library that happens to be bundled within V8. Before this users would not be able to use cppgc in their addons anyway so we could state in the beginning that if the addons use cppgc, we do not guarantee that the addons can be loaded in same major releases. For the vast majority of addons that do not use cppgc it would not matter because the new headers are just added symbols which are ABI compatible. We should be able to enforce this by making the cppgc stuff only visible when the addons explicitly request it (e.g. adding an ifdef to the headers, so the headers are still part of the tarball but addons now need to get access to what's in cppgc via an explicit define, and when they do that they are opting into being at their own risk in terms of ABI compatibility)

Reply:
Actually I took a look at the 11.3.244.8...11.6.130 changes (V8 on the main branch here and the latest V8 release) and it seems there aren't actually that many ABI breakages in cppgc - most changes are additions - compared to what happens in other V8 headers. If we maintain the current practice, which is upgrading V8 a few times (no more than once per month, I gather) before the release goes LTS, cppgc probably would not be that much of a problem (other than we would have a bit more changes to review). We do still need to pay a bit more attention to the template changes, but this also already happens to the fast API calls header (also template-heavy) and there is also growing number of inlined implementations in other headers that lead to similar problems. cppgc probably only contributes <10% to the overall ABI compat problem.

> Assuming we publish all these internal cppgc headers. Can we still update V8 in an ABI-compatible way if they make breaking changes to them (while keeping v8-cppgc.h without breaking changes)?

From what I can tell if the users do not actually use any thing from `cppgc/` it does not matter for them. And they are not supposed to use anything from `cppgc/internal` (V8 very explicitly warns against using anything there) so changes to the internal headers would matter less. The safer choice might still be that we state that we do not guarantee ABI compatibility for `cppgc` for now but might consider including it when we think it's ready.

Reply:
Opened https://bugs.chromium.org/p/v8/issues/detail?id=14062 in the upstream about cppgc ABI compatibility.

Reply:
From what I got from the v8 issue above my conclusion is that the cppgc headers should not introduce that much ABI breakages for the 6-month cadences we have, it's also only a concern for users who actually use any of the cppgc APIs, which should be niche at least for now. On the safe side we can also just state that embedder support for cppgc is experimental and it is not included in our ABI compatibility guarantee for now.